### ISSUE TITLE
Managed Data Breakpoints don't work on Windows 7 when targeting x64.

### ISSUE THREAD

**Issue by chuckries** [dotnet/coreclr#25237]
**It is important that this gets fixed for .NET Core 3.0**

As the title says, managed data breakpoints do not work on Windows 7 against an x64 target app. They appear to work when targeting x86.

Repro steps:
1. Get a machine running Windows 7
2. Install VS 2019 (either release or public preview)
3. Install .NET Core 3.0 Preview 6.
4. Create a new console app with the following code:
```cs
using System;

namespace ConsoleApp1
{
    class Test
    {
        public int One { get; set; }
        public int Two { get; set; }
    }

    class Program
    {
        static void Main(string[] args)
        {
            Test t = new Test();
            Console.WriteLine("Hello World!"); // <-- bp here
            t.One = 1;
            t.Two = 2;
        }
    }
}
```
5. set a breakpoint on the Console.WriteLine, run the app and stop at the breakpoint
6. From locals window, expand `t`
7. Right click 'One', select 'Break When Value Changes'
8. Continue the process

Expected: Debugger stops inside the property setter for One
Actual: Process continues to completion, never stops under the debugger.

This appears to be an issue with the ICorDebugProcess::SetThreadContext API, as the dr0 register is never getting written to the actual thread context, but it is returning S_OK. Therefore the debugger thinks all is well, but the data bp is never actually written into the process.

**Comment by chuckries** [comment]
cc @caslan @cshung @tommcdon

**Comment by cshung** [comment]
tl;dr - I narrowed this issue down to the fact that the WIndows 7 is dropping the data breakpoint register changes. Below is my experiment that proves that fact.

1. Checkout my [branch](https://github.com/cshung/coreclr/tree/dev/andrewau/25237-investigation), build it for debug, x64

2. Create a .NET Core 3 console application with the repro code, publish it in self contained mode and patch the coreclr with the built binaries.

3. Debug the built exe using Visual Studio with breakpoint set on the indicated line.

4. While the 1st breakpoint hit, set the data breakpoint, and then non-invasive attach WinDBG to the debuggee process and turn on inspection by running `?? coreclr!g_inspect = 1`, and then detach WinDBG to allow further execution.

5. Continue the process in Visual Studio, the process should be infinitely looping on the first loop.

6. Non-invasive attach WinDBG to the debuggee process, `pExceptionInfo->ContextRecord->Dr0` should have a non-zero value, indicating the debugger intended to stop there. The process is going to unwind out of the vectored exception handler and will return to user code using the modified context. Physically, the code run back into the Windows supporting code. Logically, the flow of control should reach the bypass buffer, which is 
`pExceptionInfo->ContextRecord->Rip`, which should also be
 `((DebuggerPatchSkip*)DebuggerController::g_controllers)->m_pSharedPatchBypassBuffer`. 
At this point of time, the single stepping flag is set, the debuggee is ready to single step through the instruction in the bypass buffer. Be careful here, do **not** detach yet, we need to set things up on the debugger side.

7. Attach a debugger to `msvsmon.exe` and break on `mscordbi!ShimProcess::HandleWin32DebugEvent`, this should capture the single stepping event, let the debugger process runs.

8. Now running `?? coreclr!g_inspect = 0` and detach the WinDBG connected to the debuggee, this should allow the debuggee to move on and single step, the breakpoint should hit on the debugger side.

9. Now inspect the debuggee again using non-invasive attach, the `r dr0` returns 0, indicating the data breakpoint register is somehow not set on the thread, but logically that thread context was set and only one instruction is executed.

**Comment by sdmaclea** [comment]
Based on @cshung analysis, this is a Windows 7 OS issue.

Also...

```After 10 years, support for Windows 7 is coming to an end on January 14, 2020.```

If this is a runtime issue the fix will not be included in a release prior to end of life for Win7.
